File: NextApp/Android/feature/tools/src/main/kotlin/jp/komehyappyo/member/next/feature/tools/DiaryFeatureModel.kt
package jp.komehyappyo.member.next.feature.tools

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import jp.komehyappyo.member.next.core.data.DiaryBackupService
import jp.komehyappyo.member.next.core.data.DiaryPhotoStore
import jp.komehyappyo.member.next.core.data.DiaryRepository
import jp.komehyappyo.member.next.core.model.Diary
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.UUID

data class DiaryUiState(
    val diaries: List<Diary> = emptyList(),
    val isLoading: Boolean = true,
    val errorMessage: String? = null,
)

class DiaryFeatureModel(
    private val repository: DiaryRepository,
    private val photoStore: DiaryPhotoStore,
    private val backupService: DiaryBackupService,
) : ViewModel() {
    private val mutableState = MutableStateFlow(DiaryUiState())
    val state: StateFlow<DiaryUiState> = mutableState

    init {
        reload()
    }

    fun reload() {
        viewModelScope.launch {
            mutableState.update { it.copy(isLoading = true, errorMessage = null) }
            repository.observeAll()
                .catch { error ->
                    mutableState.update {
                        it.copy(isLoading = false, errorMessage = error.localizedMessage)
                    }
                }
                .collect { diaries ->
                    mutableState.update { it.copy(diaries = diaries, isLoading = false) }
                }
        }
    }

    fun save(
        diary: Diary,
        newPhotoData: List<ByteArray>,
        removedPhotoReferences: Set<String>,
        onComplete: (Result<Unit>) -> Unit,
    ) {
        viewModelScope.launch {
            runCatching {
                val remaining = diary.photoUrls.filterNot(removedPhotoReferences::contains)
                require(remaining.size + newPhotoData.size <= Diary.MAXIMUM_PHOTO_COUNT) {
                    "写真は${Diary.MAXIMUM_PHOTO_COUNT}枚まで登録できます。"
                }
                val createdReferences = mutableListOf<String>()
                try {
                    withContext(Dispatchers.IO) {
                        newPhotoData.forEach { data ->
                            createdReferences += photoStore.saveJpeg(data, diary.id)
                        }
                    }
                    repository.save(
                        diary.copy(
                            photoUrls = remaining + createdReferences,
                            updatedAt = java.time.Instant.now(),
                        ),
                    )
                    withContext(Dispatchers.IO) {
                        removedPhotoReferences.forEach {
                            runCatching { photoStore.delete(it) }
                        }
                    }
                } catch (error: Throwable) {
                    withContext(Dispatchers.IO) {
                        createdReferences.forEach { runCatching { photoStore.delete(it) } }
                    }
                    throw error
                }
            }.also(onComplete)
        }
    }

    fun delete(diary: Diary, onComplete: (Result<Unit>) -> Unit) {
        viewModelScope.launch {
            runCatching {
                repository.delete(diary.id)
                withContext(Dispatchers.IO) {
                    diary.photoUrls.forEach { runCatching { photoStore.delete(it) } }
                }
            }.also(onComplete)
        }
    }

    suspend fun photoData(reference: String): ByteArray =
        withContext(Dispatchers.IO) { photoStore.load(reference) }

    fun exportBackup(onComplete: (Result<ByteArray>) -> Unit) {
        viewModelScope.launch {
            runCatching {
                withContext(Dispatchers.IO) { backupService.exportData() }
            }.also(onComplete)
        }
    }

    fun importBackup(data: ByteArray, onComplete: (Result<Int>) -> Unit) {
        viewModelScope.launch {
            runCatching {
                withContext(Dispatchers.IO) { backupService.importData(data) }
            }.also(onComplete)
        }
    }

    class Factory(
        private val repository: DiaryRepository,
        private val photoStore: DiaryPhotoStore,
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T =
            DiaryFeatureModel(
                repository,
                photoStore,
                DiaryBackupService(repository, photoStore),
            ) as T
    }
}

File: NextApp/iOS/OrgPortalNext/Feature/Tools/DiaryFeatureModel.swift
import DataLayer
import Foundation
import Model

@MainActor
public final class DiaryFeatureModel: ObservableObject {
    @Published public private(set) var diaries: [Diary] = []
    @Published public private(set) var isLoading = false
    @Published public private(set) var errorMessage: String?

    private let repository: DiaryRepository
    private let photoStore: DiaryPhotoStoring
    private let backupService: DiaryBackupService

    public init(
        repository: DiaryRepository,
        photoStore: DiaryPhotoStoring
    ) {
        self.repository = repository
        self.photoStore = photoStore
        backupService = DiaryBackupService(
            repository: repository,
            photoStore: photoStore
        )
    }

    public func load() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }
        do {
            diaries = try await repository.fetchAll()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func save(
        _ diary: Diary,
        newPhotoData: [Data],
        removedPhotoReferences: Set<String>
    ) async throws {
        let remainingReferences = diary.photoUrls.filter {
            !removedPhotoReferences.contains($0)
        }
        guard remainingReferences.count + newPhotoData.count <= Diary.maximumPhotoCount else {
            throw DiaryValidationError.tooManyPhotos
        }

        var createdReferences: [String] = []
        do {
            for data in newPhotoData {
                createdReferences.append(
                    try photoStore.saveJPEGData(data, diaryId: diary.id)
                )
            }
            var value = diary
            value.photoUrls = remainingReferences + createdReferences
            value.updatedAt = .now
            try await repository.save(value)
            for reference in removedPhotoReferences {
                try? photoStore.delete(reference: reference)
            }
            await load()
        } catch {
            for reference in createdReferences {
                try? photoStore.delete(reference: reference)
            }
            throw error
        }
    }

    public func delete(_ diary: Diary) async {
        do {
            try await repository.delete(id: diary.id)
            for reference in diary.photoUrls {
                try? photoStore.delete(reference: reference)
            }
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func photoData(reference: String) throws -> Data {
        try photoStore.loadData(reference: reference)
    }

    public func exportBackup() async throws -> Data {
        try await backupService.exportData()
    }

    @discardableResult
    public func importBackup(_ data: Data) async throws -> Int {
        let count = try await backupService.importData(data)
        await load()
        return count
    }
}
